File: c/src/zmqinterface.c
#include "zmqinterface.h"

#include "zmq.h"

struct ZMQInterfaceCDT {
   void *context;
   void *server;
   void *client;
};

ZMQInterface newZMQInterface(int port_recv, int port_send) {
   ZMQInterface iface = newBlock(ZMQInterface);
   int rc = 0;

   iface->context = zmq_ctx_new();
   iface->server = zmq_socket(iface->context, ZMQ_REP);
   iface->client = zmq_socket(iface->context, ZMQ_REQ);

   do {
      string tmp = concat("tcp://*:", integerToString(port_recv));
      rc = zmq_bind(iface->server, (char *)tmp);
      port_recv++;

#ifdef ZMQDEBUG
      if (rc == -1) {
         fprintf(stderr,
                 "DEBUG: ZMQInterface failed to open Server. Port %d occupied, "
                 "trying port %d.\n",
                 port_recv - 1, port_recv);
      } else {
         fprintf(stderr,
                 "DEBUG: ZMQInterface successfully opened Server at port %d\n",
                 port_recv - 1);
      }
#endif

   } while (rc == -1);

   string tmp = concat("tcp://localhost:", integerToString(port_send));
   rc = zmq_connect(iface->client, (char *)tmp);

#ifdef ZMQDEBUG
   if (rc == -1) {
      fprintf(stderr, "DEBUG: ZMQInterface failed to connect to %s.\n", tmp);
   } else {
      fprintf(
          stderr,
          "DEBUG: ZMQInterface successfully changed client-connection to %s.\n",
          tmp);
   }
#endif

   return iface;
}

void updateClientPort(ZMQInterface iface, int port_send) {
   zmq_close(iface->client);
   iface->client = zmq_socket(iface->context, ZMQ_REQ);
   string tmp = concat("tcp://localhost:", integerToString(port_send));
   int rc = zmq_connect(iface->client, (char *)tmp);

#ifdef ZMQDEBUG
   if (rc == -1) {
      fprintf(stderr, "DEBUG: ZMQInterface failed to connect to %s.\n", tmp);
   } else {
      fprintf(stderr, "DEBUG: ZMQInterface successfully connected to %s.\n",
              tmp);
   }
#endif
}

// FIXME: Need to call @end of program
void freeZMQInterface(ZMQInterface iface) {
   zmq_close(iface->client);
   zmq_close(iface->server);
   zmq_ctx_destroy(iface->context);
   free(iface);

#ifdef ZMQDEBUG
   fprintf(stderr, "DEBUG: ZMQInterface-instance destroyed.\n");
#endif
}

void sendString(ZMQInterface iface, string str, bool last) {
   if (last) {
      zmq_send(iface->client, (byte *)str, stringLength(str), 0);
      char reply[2];
      zmq_recv(iface->client, reply, 2, 0);

#ifdef ZMQDEBUG
      if (stringEqual((string)reply, "0")) {
         fprintf(stderr,
                 "DEBUG: ZMQInterface successfully send data to Server.\n");
      } else {
         fprintf(stderr,
                 "DEBUG: Something went wrong, while sending data to Server. "
                 "Server returned ZMQinterface-Error #%s.\n",
                 reply);
      }
#endif

   } else {
      zmq_send(iface->client, (byte *)str, stringLength(str), ZMQ_SNDMORE);
   }
}

string recvString(ZMQInterface iface, int buffer_len) {
   string str = newArray(buffer_len, char);

   zmq_recv(iface->server, str, buffer_len, 0);
   return str;
}

void sendIntArray(ZMQInterface iface, int *arr, int len, bool last) {
   if (last) {
      zmq_send(iface->client, (byte *)arr, len * sizeof(int), 0);
      char reply[2];
      zmq_recv(iface->client, reply, 2, 0);

#ifdef ZMQDEBUG
      if (stringEqual((string)reply, "0")) {
         fprintf(stderr,
                 "DEBUG: ZMQInterface successfully send data to Server.\n");
      } else {
         fprintf(stderr,
                 "DEBUG: Something went wrong, while sending data to Server. "
                 "Server returned ZMQinterface-Error #%s.\n",
                 reply);
      }
#endif

   } else {
      zmq_send(iface->client, (byte *)arr, len * sizeof(int), ZMQ_SNDMORE);
   }
}

void recvIntArray(ZMQInterface iface, int **arr) {
   int len = 0;
   zmq_recv(iface->server, (byte *)&len, sizeof(int), 0);
   *arr = newArray(len, int);
   zmq_recv(iface->server, (byte *)(*arr), len * sizeof(int), 0);
}

void sendConfirmationReply(ZMQInterface iface) {
   zmq_send(iface->server, "0", 1, 0);
#ifdef ZMQDEBUG
   fprintf(stderr,
           "DEBUG: ZMQInterface successfully received data from Client. "
           "Sending confirmation to Client.\n");
#endif
}

File: java/src/stanford/spl/ZMQInterface.java
/*
 * File: ZMQInterface.java 
 * ----------------------- 
 * This file implements the ZMQ-Interface for the JBE for binary transport 
 * of data to the C FrontEnd.
 */

/*************************************************************************/
/* Stanford Portable Library                                             */
/* Copyright (C) 2013 by Eric Roberts <eroberts@cs.stanford.edu>         */
/*                                                                       */
/* This program is free software: you can redistribute it and/or modify  */
/* it under the terms of the GNU General Public License as published by  */
/* the Free Software Foundation, either version 3 of the License, or     */
/* (at your option) any later version.                                   */
/*                                                                       */
/* This program is distributed in the hope that it will be useful,       */
/* but WITHOUT ANY WARRANTY; without even the implied warranty of        */
/* MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the         */
/* GNU General Public License for more details.                          */
/*                                                                       */
/* You should have received a copy of the GNU General Public License     */
/* along with this program.  If not, see <http://www.gnu.org/licenses/>. */
/*************************************************************************/

package stanford.spl;

import org.zeromq.ZMQ;
import org.zeromq.ZMQException;
import org.zeromq.SocketType;
import org.zeromq.ZContext;

public class ZMQInterface {

   private boolean debug;
   private ZContext context;
   private ZMQ.Socket server;
   private ZMQ.Socket client;
   private int server_port;

   public ZMQInterface(boolean debug, int port_recv, int port_send) {
      this.debug = debug;
      context = new ZContext();
      server = context.createSocket(SocketType.REP);
      client = context.createSocket(SocketType.REQ);
      boolean bound = false;

      while (!bound) {
         try {
            server.bind("tcp://*:" + port_recv);
            bound = true;
            server_port = port_recv;
            if (this.debug)
               System.err.println(
                     "JavaBackEnd: ZMQInterface: Server listening at port " + port_recv + ".");
         } catch (ZMQException ex) {
            if (this.debug)
               System.err.println("JavaBackEnd: ZMQInterface: Port " + port_recv
                     + " is already occupied, trying port " + (port_recv + 1) + ".");
            port_recv++;
         }
      }

      try {
         client.connect("tcp://localhost:" + port_send);
         if (this.debug)
            System.err.println("JavaBackEnd: ZMQInterface: Client now sends messages to port "
                  + port_send + ".");
      } catch (ZMQException ex) {
         if (this.debug)
            System.err.println(
                  "JavaBackEnd: ZMQInterface: Client could not connect to port " + port_send + ".");
      }

      if (this.debug)
         System.err.println("JavaBackEnd: ZMQInterface: Initialized ZMQInterface.");

   }

   public int getServerPort() {
      return server_port;
   }


   // public void sendString(String str) {
   // client.send(str.getBytes(ZMQ.CHARSET), 0);
   // }

   // public String recvString() {
   // return server.recvStr();
   // }

   public void sendIntArray(int[] arr, boolean last) {
      byte[] len_bin = new byte[4];
      byte[] arr_bin = new byte[arr.length << 2];
      int len = arr.length;

      len_bin[0] = (byte) ((len >>> 0) & 0xff);
      len_bin[1] = (byte) ((len >>> 8) & 0xff);
      len_bin[2] = (byte) ((len >>> 16) & 0xff);
      len_bin[3] = (byte) ((len >>> 24) & 0xff);
      client.send(len_bin, ZMQ.SNDMORE);

      for (int i = 0; i < arr.length; i++) {
         int x = arr[i];
         int j = i << 2;
         arr_bin[j++] = (byte) ((x >>> 0) & 0xff);
         arr_bin[j++] = (byte) ((x >>> 8) & 0xff);
         arr_bin[j++] = (byte) ((x >>> 16) & 0xff);
         arr_bin[j++] = (byte) ((x >>> 24) & 0xff);
      }
      if (last) {
         client.send(arr_bin, 0);
         String reply = new String(client.recv(0), ZMQ.CHARSET);
         if (this.debug) {
            if(reply.equals("0")) {
               System.err.println("JavaBackEnd: ZMQInterface: Successfully send data to Server.");
            } else {
               System.err.println("JavaBackEnd: ZMQInterface: Something went wrong, while sending data to " +
                  "Server. Server returned ZMQinterface-Error #" + reply);
            }
         }
      } else {
         client.send(arr_bin, ZMQ.SNDMORE);
      }
   }

   public void sendConfirmationReply() {
      String msg = "0";
      server.send(msg.getBytes(ZMQ.CHARSET), 0);
      if (this.debug)
         System.err.println(
               "JavaBackEnd: ZMQInterface: Successfully received data from Client. Sending confirmation to Client.");
   }

   public int[] recvIntArray() {
      byte[] arr_bin = server.recv(0);
      int[] arr = new int[arr_bin.length >>> 2];

      for (int i = 0; i < arr.length; i++) {
         int j = i << 2;
         int x = 0;
         x += (arr_bin[j++] & 0xff) << 0;
         x += (arr_bin[j++] & 0xff) << 8;
         x += (arr_bin[j++] & 0xff) << 16;
         x += (arr_bin[j++] & 0xff) << 24;
         arr[i] = x;
      }
      return arr;
   }

   public void destroy() {
      context.close();
      if (this.debug)
         System.err.println("JavaBackEnd: ZMQInterface: Destroyed ZMQInterface.");
   }

}
